src/Shared/LayoutRenderers/AspNetRequestDurationLayoutRenderer.cs

```diff
@@ -0,0 +1,186 @@
+                {
+                    RenderDurationMs(builder, duration.Value.TotalMilliseconds);
+                }
+                else
+                {
+                    builder.Append(FormatDuration(duration.Value));
+                }
+            }
+#endif
+        }
+
+        private void RenderDurationMs(StringBuilder builder, double durationMs)
+        {
+            if (ReferenceEquals(Culture, System.Globalization.CultureInfo.InvariantCulture))
+            {
+                var truncateMs = (long)durationMs;
+                if (DurationMsFormat != null && truncateMs >= 0 && truncateMs <= DurationMsFormat.Length)
+                {
+                    builder.Append(DurationMsFormat[truncateMs]);
+                }
+                else
+                {
+                    builder.Append(truncateMs);
+                }
+                var preciseMs = (int)((durationMs - truncateMs) * 1000.0);
+                if (preciseMs > 0)
+                {
+                    builder.Append('.');
+                    if (preciseMs < 100)
+                        builder.Append('0');
+                    if (preciseMs < 10)
+                        builder.Append('0');
+
+                    if (DurationMsFormat != null && preciseMs <= DurationMsFormat.Length)
+                    {
+                        builder.Append(DurationMsFormat[preciseMs]);
+                    }
+                    else
+                    {
+                        builder.Append(preciseMs);
+                    }
+                }
+                else
+                {
+                    builder.Append(".0");
+                }
+            }
+            else
+            {
+                builder.Append(durationMs.ToString("0.###", Culture));
+            }
+        }
+
+        string FormatDuration(TimeSpan duration)
+        {
+#if !ASP_NET_CORE
+            if (string.IsNullOrEmpty(_formatString))
+                return duration.ToString();
+            else
+                return string.Format(_formatString, duration);
+#else
+            return duration.ToString(Format, Culture);
+#endif
+        }
+
+#if !ASP_NET_CORE
+        private static TimeSpan? GetDuration(DateTime contextTimestamp)
+        {
+            if (contextTimestamp > DateTime.MinValue)
+            {
+                if (contextTimestamp.Kind == DateTimeKind.Local)
+                    return DateTime.Now - contextTimestamp;
+                else if (contextTimestamp.Kind == DateTimeKind.Unspecified)
+                    return DateTime.UtcNow - new DateTime(contextTimestamp.Ticks, DateTimeKind.Utc);
+                else
+                    return DateTime.UtcNow - contextTimestamp.ToUniversalTime();
+            }
+            else
+            {
+                return default(TimeSpan?);
+            }
+        }
+#endif
+
+#if ASP_NET_CORE && !ASP_NET_CORE2
+        private static TimeSpan? GetDuration(System.Diagnostics.Activity activity)
```

Comment:
I'm not sure if all the paths has unit tests. I would recommend to add them

Reply:
Think I will extract AspNetRequestDurationLayoutRenderer into its own pull-request. Where more attention can be given to test-coverage. I just wanted to see how easy it was to use NLog for request-logging.

Reply:
good idea!